### PR TITLE
BNO05 IMU ROS integration

### PR DESCRIPTION
**Pull request by jheidegger**
using last years wrapper for the BNO055 to create a ROS node as a part of the I2C node. It will have its own topic imu_data. Created a message type for the IMU. Needs hardware verification.

### CODE REVIEW

File: ros/src/i2c_com/scripts/BNO055.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+        self._data['temp'] = temp
+
+        return True
+
+    def get_calibration(self):
+        return self._bno.get_calibration()
+
+    def reset_calibration(self):
+        cal_array_original = self.get_calibration()
+        self._bno.set_calibration(self._bno.get_calibration())
```
Comment (bmaxfie) [inline ros/src/i2c_com/scripts/BNO055.py]: Why call get_calibration() again? Just pass cal_array_original right?

File: ros/src/i2c_com/scripts/BNO055.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    def main():
+        sensor = BNO055() # Default I2C bus is 1 (Raspberry Pi 3)
+
+        # We must initialize the sensor before reading it
+        if not sensor:
+                print "Sensor could not be initialized"
+                exit(1)
+
+        # We have to read values from sensor to update pressure and temperature
+        #if not sensor.read():
```
Comment (bmaxfie) [inline ros/src/i2c_com/scripts/BNO055.py]: Remove deprecated code.

File: ros/src/i2c_com/scripts/i2c.py [hunk 3]
```diff
@@ -1,17 +1,31 @@
...
 
 if __name__ == "__main__":
   rospy.init_node('i2c_node')
-  pub = rospy.Publisher('i2c_rx', i2c_msg,
-      queue_size= 100)
-  sub = rospy.Subscriber('i2c_tx', i2c_msg,
-      message_received)
+  imu_pub = rospy.Publisher('imu_data', imu_msg,
+	qeue_size = 100)
+  imu_pub.rate(50)
```
Comment (bmaxfie) [inline ros/src/i2c_com/scripts/i2c.py]: Have rate as constant (Put 50 at top as constant), so we can more easily update it.

File: ros/src/i2c_com/scripts/i2c.py [hunk 4]
```diff
@@ -1,17 +1,31 @@
...
 import rospy
+import time
+import smbus
+from BNO055 import BNO055
 from shared_msgs.msg import i2c_msg
+from shared_msgs.msg import imu_msg
+
+
 
 def message_received(msg):
```
Comment (bmaxfie) [inline ros/src/i2c_com/scripts/i2c.py:11]: Remove unused code

File: ros/src/imu/imu.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
+#! /usr/bin/python
+import rospy
+#TODO make imu_message
+
+def message_received(msg):
+  # This runs on a seperate thread from the pub
+  pass
+
+if __name__ == "__main__":
```
Comment (bmaxfie) [inline ros/src/imu/imu.py]: Remove this imu folder.

File: ros/src/i2c_com/scripts/i2c.py [hunk 6]
```diff
@@ -1,17 +1,31 @@
...
+  imu_pub = rospy.Publisher('imu_data', imu_msg,
+	qeue_size = 100)
+  imu_pub.rate(50) 
+  imu = BNO055()
+  if not sensor:
+    raise RuntimeError('Failed to initialize BNO055!')
+  while not rospy.is_shutdown():
+    if imu.update():
+      out_message = imu_msg()
+      out_message.gyro = {imu.gyro_x, imu.gyro_y, imu.gyro_z}
```
Comment (bmaxfie) [inline ros/src/i2c_com/scripts/i2c.py]: Don't you need to make these gyro_*/acceleration_* calls functions unless they're labeled as @property right?

**Review by bmaxfie — CHANGES_REQUESTED** [review]

File: ros/src/i2c_com/scripts/i2c.py (hunk 6 above)
Comment (jheidegger) [reply ros/src/i2c_com/scripts/i2c.py]: changed to the function call for greater flexibility down the line

File: ros/src/i2c_com/scripts/BNO055.py (hunk 1 above)
Comment (jheidegger) [reply ros/src/i2c_com/scripts/BNO055.py]: the get calibration call is to the adafruit library. It is calling the constant hard coded and defined in there. This code is for if we do custom calibration or the calibration data is corrupted we want to get a log of what the calibration data was before we reset it to the library standard

File: ros/src/i2c_com/scripts/BNO055.py (hunk 2 above)
Comment (jheidegger) [reply ros/src/i2c_com/scripts/BNO055.py]: I believe this code is still useful for testing the system in the future.

**Comment by jheidegger** [comment]
Ready to be merged

**Comment by jheidegger** [comment]
@bmaxfie using this pull request to continue as to avoid branch conflicts. Temperature sensor is added to ros. I have a few questions on polling rates  that we may have to split sensors into their own nodes not sure  just yet. The main issue is if they have different rates at which they send data which rate do we send  the data at. So thats just what i'm working through now. I will be busy the rest of the weekend.

**Comment by bmaxfie** [comment]
Let's hope we get lucky and the update rates are multiples/factors of each other. Then we can have one node that can update both at the smaller interval's sleep rate.

File: ros/src/i2c_com/scripts/i2c.py [hunk 7]
```diff
@@ -1,17 +1,38 @@
...
+      out_message.accel = {imu.acceleration_x, imu.acceleration_y, imu.acceleration_z}
+      imu_pub.publish(out_message)
+      # read the temp in C and send it to a ros
+      temp_message = temp_msg()
+      temp_message.temperature = {temp.temperature()}
+      temp_pub.publish(temp_message)
+      r.sleep()
 
-  # TODO: I2C related activities
   rospy.spin()
```
Comment (bmaxfie) [inline ros/src/i2c_com/scripts/i2c.py]: Pretty sure you should remove rospy.spin() if it's only going to get called if ROS is shutdown.

**Comment by jheidegger** [comment]
@bmaxfie added the depth sensor for review

**Comment by jheidegger** [comment]
@bmaxfie  pull request updated to merge to ROS

File: ros/src/i2c_com/scripts/BNO055.py [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+        if test_array[6] != 0:
+            return False
+
+        return True
+
+if __name__ == '__main__':
+    #BNO055().main()
+    import time
+
+    def main():
```
Comment (bmaxfie) [inline ros/src/i2c_com/scripts/BNO055.py:136]: Not a big fan of the internal main function but whatever.

File: ros/src/i2c_com/scripts/TYS01.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    _TSYS01_READ        = 0x00
+
+    def __init__(self, bus=1):
+        # Degrees C
+        self._temperature = 0
+        self._k = []
+
+        try:
+            self._bus = smbus.SMBus(bus)
+        except:
```
Comment (bmaxfie) [inline ros/src/i2c_com/scripts/TYS01.py:26]: Translate this caught exception into a Runtime exception or something that all of the sensor classes will throw. I explain the reason for this in another comment.

File: ros/src/i2c_com/scripts/i2c.py [hunk 10]
```diff
@@ -1,17 +1,48 @@
...
+from TYS01 import TSYS01
+from ms5837 import MS5837
+from shared_msgs.msg import imu_msg
+from shared_msgs.msg import temp_msg
+from shared_msgs.msg import depth_msg
+
+IMU_PUB_RATE = 50
+TEMP_RATIO_RATE = 4
 
 def message_received(msg):
```
Comment (bmaxfie) [inline ros/src/i2c_com/scripts/i2c.py:11]: You got rid of the code inside but not the unused function definition?

File: ros/src/i2c_com/scripts/i2c.py [hunk 11]
```diff
@@ -1,17 +1,48 @@
...
-  sub = rospy.Subscriber('i2c_tx', i2c_msg,
-      message_received)
-
-  # TODO: I2C related activities
-  rospy.spin()
+  imu_pub = rospy.Publisher('imu_data', imu_msg, qeue_size = 100)
+  imu_pub.rate(IMU_PUB_RATE)
+  temp_pub = rospy.Publisher('temp_data', temp_msg, qeue_size = 100)
+  depth_pub = rospy.Publisher('depth_data', depth_msg, qeue_size = 100)
+  imu = BNO055()
```
Comment (bmaxfie) [inline ros/src/i2c_com/scripts/i2c.py]: Add catches to these sensor constructor calls that will, if they throw a Runtime Exception or something like that, you replace the sensor with a mock (and a print out explaining that the sensor failed to be initialized).

Furthermore, you will have to make mock classes, not that difficult. Just fill the interfaces and return a sine wave or something like that.

File: ros/src/i2c_com/scripts/i2c.py [hunk 12]
```diff
@@ -1,17 +1,48 @@
...
+  temp = TYS01()
+  depth = ms5837()
+  temp.init()
```
Comment (bmaxfie) [inline ros/src/i2c_com/scripts/i2c.py]: Kinda odd that this is the only class with an init function? Consider merging into the constructor?

File: ros/src/i2c_com/scripts/i2c.py [hunk 13]
```diff
@@ -1,17 +1,48 @@
...
+      imu_pub.publish(out_message)
+      # read the temp in C and send it to a ros
+      if(temp_sleep % TEMP_RATIO_RATE == 0) {
+          temp_message = temp_msg()
+          temp_message.temperature = {temp.temperature()}
+          temp_pub.publish(temp_message)
+      }
+      depth.update()
+      depth_message = depth_msg()
+      depth_msg.depth = {depth.depth()}
```
Comment (bmaxfie) [inline ros/src/i2c_com/scripts/i2c.py]: depth_message.depth = {depth.depth()}

I believe is what you mean.

File: ros/src/i2c_com/scripts/i2c.py (hunk 11 above)
Comment (bmaxfie) [reply ros/src/i2c_com/scripts/i2c.py]: Also not a big fan of just calling the sensors imu, temp, and depth. I'd consider appending at least _sens or something at the end.

File: ros/src/i2c_com/scripts/i2c.py (hunk 13 above)
Comment (bmaxfie) [reply ros/src/i2c_com/scripts/i2c.py]: Also, are you sure enclosing the values in {} works with the data types of these messages?

File: ros/src/i2c_com/scripts/ms5837.py [hunk 14]
```diff
@@ -0,0 +1,304 @@
...
+    _MS5837_CONVERT_D1_256   = 0x40
+    _MS5837_CONVERT_D2_256   = 0x50
+    
+    def __init__(self, model=MODEL_30BA, bus=1):
+        self._model = model
+        
+        try:
+            self._bus = smbus.SMBus(bus)
+        except:
+            print("Bus %d is not available.") % bus
```
Comment (bmaxfie) [inline ros/src/i2c_com/scripts/ms5837.py:56]: Same thing as temp sensor. throw exception to be caught later.

**Review by bmaxfie — CHANGES_REQUESTED** [review]

**Review by bmaxfie — APPROVED** [review]

**Pull request merged**